pageserver/benches/bench_walredo.rs

```diff
@@ -74,114 +75,83 @@ fn redo_scenarios(c: &mut Criterion) {
-                            if work_rx.lock().unwrap().recv().is_err() {
-                                break;
-                            }
-
-                            let input = Some(input_factory());
-
-                            barrier.wait();
-
-                            execute_all(input, handle, &manager).unwrap();
-
-                            barrier.wait();
-                        }
+    assert_ne!(requesters, 0);
+
+    let rt = tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    let cancel = CancellationToken::new();
+
+    let barrier = Arc::new(tokio::sync::Barrier::new(requesters + 1));
+
+    let jhs = (0..requesters)
+        .map(|_| {
+            let manager = manager.clone();
+            let barrier = barrier.clone();
+            let cancel = cancel.clone();
+            rt.spawn(async move {
+                let work_loop = async move {
+                    loop {
+                        let input = input_factory();
+                        barrier.wait().await;
+                        let page = input.execute(&manager).await.unwrap();
+                        assert_eq!(page.remaining(), 8192);
+                        barrier.wait().await;
                     }
-                })
-            })
-            .collect::<Vec<_>>();
-
-        let _jhs = JoinOnDrop(jhs);
-
-        b.iter_batched(
-            || {
-                for _ in 0..threads {
-                    work_tx.send(()).unwrap()
+                };
+                tokio::select! {
+                    _ = work_loop => {},
+                    _ = cancel.cancelled() => { }
                 }
-            },
-            |()| {
-                // start the work
-                barrier.wait();
-
+            })
+        })
+        .collect::<Vec<_>>();
+
+    b.iter_batched(
+        || (),
+        |()| {
+            // start the work
+            rt.block_on(async {
+                barrier.wait().await;
                 // wait for work to complete
-                barrier.wait();
-            },
-            criterion::BatchSize::PerIteration,
-        );
+                barrier.wait().await;
+            });
+        },
+        criterion::BatchSize::PerIteration,
+    );
 
-        drop(work_tx);
-    }
-}
+    cancel.cancel();
 
-struct JoinOnDrop(Vec<std::thread::JoinHandle<()>>);
-
-impl Drop for JoinOnDrop {
-    // it's not really needless because we want join all then check for panicks
-    #[allow(clippy::needless_collect)]
-    fn drop(&mut self) {
-        // first join all
-        let results = self.0.drain(..).map(|jh| jh.join()).collect::<Vec<_>>();
-        // then check the results; panicking here is not great, but it does get the message across
-        // to the user, and sets an exit value.
-        results.into_iter().try_for_each(|res| res).unwrap();
-    }
+    let jhs = JoinOnDrop(jhs);
+    rt.block_on(jhs.join_all());
```

Comment:
```suggestion
    futures::future::try_join_all(jhs).await.unwrap();
```

Though, why not joinset?

Reply:
Because I didn't want to do more changes than necessary, but yeah, let's do the switch to JoinSet